### PR TITLE
Refactor streaming's filtering logic & improve documentation

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
This is a re-open of #26172 which was closed when @ClearlyClaire merged #26159. See #26172 for details.

### CODE REVIEW

File: streaming/index.js [hunk 1]
```diff
@@ -622,41 +622,57 @@ const startServer = async () => {
...
+    // The listener used to process each message off the redis subscription,
+    // message here is an object with an `event` and `payload` property. Some
+    // events also include a queued_at value, but this is being removed shortly.
+    const listener = message => {
+      const { event, payload } = message;
+
+      // Streaming only needs to apply filtering to some channels and only to
+      // some events. This is because majority of the filtering happens on the
+      // Ruby on Rails (mastodon-web) side when producing the event for
+      // streaming.
```
Comment (ThisIsMissEm) [inline streaming/index.js:642]:
```suggestion
      // Ruby on Rails side when producing the event for streaming.
```

File: streaming/index.js [hunk 2]
```diff
@@ -683,35 +699,56 @@ const startServer = async () => {
...
+          // Handling blocks & mutes and domain blocks: If one of those applies,
+          // then we don't transmit the payload of the event to the client
           if (values[0].rows.length > 0 || (accountDomain && values[1].rows.length > 0)) {
             return;
           }
 
-          if (!payload.filtered && !req.cachedFilters) {
+          // If the payload already contains the `filtered` property, it means
+          // that mastodon-web has applied the filters on the ruby on rails side,
+          // as such, we don't need to construct or apply the filters in streaming:
```
Comment (ThisIsMissEm) [inline streaming/index.js:712]:
```suggestion
          // If the payload already contains the `filtered` property, it means
          // that filtering has been applied the on the ruby on rails side, as 
          // such, we don't need to construct or apply the filters in streaming:
```

File: streaming/index.js [hunk 3]
```diff
@@ -683,35 +698,56 @@ const startServer = async () => {
...
 
+          // Handling blocks & mutes and domain blocks: If one of those applies,
+          // then we don't transmit the payload of the event to the client
           if (values[0].rows.length > 0 || (accountDomain && values[1].rows.length > 0)) {
             return;
           }
 
-          if (!payload.filtered && !req.cachedFilters) {
+          // If the payload already contains the `filtered` property, it means
+          // that filtering has been applied the on the ruby on rails side, as
```
Comment (ThisIsMissEm) [inline streaming/index.js:710]:
```suggestion
          // that filtering has been applied on the ruby on rails side, as 
```

File: streaming/index.js [hunk 4]
```diff
@@ -731,34 +767,56 @@ const startServer = async () => {
...
-          // Check filters
-          if (req.cachedFilters && !payload.filtered) {
-            const mutatedPayload = { ...payload };
+          // Apply cachedFilters against the payload, constructing a
+          // `filter_results` array of FilterResult entities
+          if (req.cachedFilters) {
             const status = payload;
-            const searchContent = ([status.spoiler_text || '', status.content].concat((status.poll && status.poll.options) ? status.poll.options.map(option => option.title) : [])).concat(status.media_attachments.map(att => att.description)).join('\n\n').replace(/<br\s*\/?>/g, '\n').replace(/<\/p><p>/g, '\n\n');
-            const searchIndex = JSDOM.fragment(searchContent).textContent;
+            // TODO: Calculate searchableContent in mastodon-web / Rails:
```
Comment (ThisIsMissEm) [inline streaming/index.js:774]:
```suggestion
            // TODO: Calculate searchableContent in Ruby on Rails:
```

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire / @renchap I've applied the comments from #26172 in the following commits: https://github.com/mastodon/mastodon/pull/26213/files/3ebd078e5be030bffe9721f74ee60e42ca34faec..eb169b7c663023a4b16e2cd5333c94194291ef71

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**